rust/gui-client/docs/process_split.md

```diff
@@ -1,34 +1,60 @@
+- Creating the tunnel interface
+- Adding / removing IP routes
+
+Anything that can work with normal user privileges should run in the non-privileged GUI process
+
+GUI process:
+
+- All Tauri code
+- `get_system_default_resolvers`
+- Listening for DNS / network changes
+
+Ambiguous:
+
+- The DNS stub resolver doesn't actually need privileges to do its work. But since it gets packets from the tunnel interface, and it's already inside connlib, and moving it would require extra IPC round-trips, it will stay inside the privileged tunnel process.
 
 ## Decisions
 
+- The CLI Client will not be split for now, it keep the same CLI interface but may be refactored.
 - The CLI and GUI client packages will have a `conflicts` tag, at least for now.
 - Eventually, Clients and Gateways will be allowed to run on the same host, so avoid conflicts there
 - If the service account token exists, we are in service mode and won't accept connections from a GUI client, even if the token is invalid.
+- The GUI stores tokens from interactive auth in a secure keyring provided by the desktop environment. (e.g. gnome-keyring) I assume the tunnel running as root with no desktop env, doesn't have access to that kind of security, since it starts without a password.
+- The GUI and privileged tunnel will live in separate binaries in case we need to set capabilities on the tunnel binary. This will also reduce the amount of code mapped into executable address space of the privileged tunnel process.
 
 ## Binaries
 
 I gave them new names to clarify.
 
-1. `firezoned` - A daemon that runs the tunnel and has elevated privilege
-2. `firezonectl` - A CLI program that runs unprivileged and exits once each operation is complete, like `systemctl` or `nmcli`
+1. `firezone-client-tunnel` - A daemon that runs the tunnel, listens for commands from the GUI over IPC, and has elevated privilege. This is a systemd service on Linux and a Windows service on Windows.
+2. `firezone` - The same daemon binary, but running as a standalone CLI / systemd service instead of listening for GUI commands
 3. `firezone-gui` - An unprivileged GUI program that keeps an icon in the system tray / notification center and is similar to `firezonectl`
 
+## Servers
+
+`firezone` (or `firezone-client-tunnel` with special CLI flags) runs all the time as a systemd service. It does not listen for GUI commands and only uses service tokens.
+
 ## Docker
 
-`firezoned` runs all the time. You can do `docker exec` to run `firezonectl` within the container.
+`firezone-client-tunnel` runs all the time and does not take interactive commands.
 
 ## Desktop
 
-`firezoned` is started as a systemd service. It tries to find a service account token and use that. If there is no service account token, it waits for a GUI to connect over IPC.
+`firezone-client-tunnel` runs all the time as a systemd service. It tries to find a service account token and use that. If there is no service account token, it waits for a GUI to connect over IPC.
```

Comment:
Why not just one binary, and if it finds a token (ENV or file) it is headless, and if it doesn't it tries to listen for IPC control by a GUI?

I think it should be up to systemd or the launching environment to set the token.

Reply:
I can do that, it's just hard to imagine a customer wanting a chance of the GUI starting up and saying, "I can't connect to the IPC server, either it's not running or it's in service account token mode"

And I won't have the token in-line in the systemd service, the token will go in `/etc` and the systemd service will be in `/usr/lib/systemd/system` so we can update it without parsing the file to grab the token.

Reply:
Sounds good, we use EnvironmentFile I believe to do this with Firezone 0.4 or something.